src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -726,86 +753,112 @@ DatabaseShardImp::onChildrenStopped()
 }
 
 void
-DatabaseShardImp::import(Database& source)
+DatabaseShardImp::importDatabase(Database& source)
 {
+    std::lock_guard lock(mutex_);
+    assert(init_);
+
+    // Only the application local node store can be imported
+    if (&source != &app_.getNodeStore())
     {
-        std::lock_guard lock(mutex_);
-        assert(init_);
+        assert(false);
+        JLOG(j_.error()) << "invalid source database";
+        return;
+    }
 
-        // Only the application local node store can be imported
-        if (&source != &app_.getNodeStore())
-        {
-            assert(false);
-            JLOG(j_.error()) << "invalid source database";
+    // Run the lengthy node store import process in the background
+    // on a dedicated thread.
+    databaseImporter_ = std::thread([this] { doImportDatabase(); });
+}
+
+void
+DatabaseShardImp::doImportDatabase()
```

Comment:
I'm wondering how this import function interacts with online delete rotation. There are a lot of error handling, e.g. checking that the expected ledgers for a shard are found, etc.

Imagine a situation where the import is running on shard X, and 95% of the way through, online delete wipes out all of the ledgers from X. The process aborts and cleans up nicely, but we've wasted a _lot_ of time and resources doing almost all of the work, only to throw it away.

I'm wondering if we can do better.

One way I can think of that might work better is to run the import backwards - from `latestIndex` to `earliestIndex`. Later indexes intuitively seem less likely to get deleted than earlier indexes (because at least an entire `online_delete`'s worth of ledgers will have to be validated before they can be deleted, whereas an early index could be deleted after the very next validated ledger). And if you do get to a point where the ledgers have been validated, it's a safe bet that _all_ of the earlier ledgers have also been validated.

Another possibility is to have some kind of interaction between the database import and `SHAMapStore`, sort of like the `can_delete` value that prevents the deletion from running while the shard is being imported.

My intention with this comment is to start a discussion, not necessarily to give instruction over changes to make. I may be completely overestimating the risk or severity of the problem, or whether there's a problem at all.

Reply:
Good point. I like the second approach.

Reply:
I know I'm coming in late on this, but you really could just disallow online delete when importing from the nodestore. If the server comes up and is trying to import from the nodestore, but sees online delete is configured in the config, just throw an error with a message saying "please disable online delete to import from the nodestore" or some such, and shut down. I see you've already wrote code to handle the synchronization, but we could probably side step that whole thing if you want to or find that preferable.

Reply:
@cjcobb23 The problem with disallowing import and online delete entirely is that if I have online delete configured and a significant amount of history (because I'm using `can_delete` or just configured a large number of ledgers), then I have to throw that all away and start over to do an import, which seems to defeat the purpose of this functionality.

Reply:
@ximinez does turning off online delete invalidate your data that was stored when online delete was on?

Reply:
@cjcobb23 Basically, yes. The directory structure is significantly different, amongst other things. 

Maybe a compromise would be to throw if online delete is configured and either `can_delete` is false or the `can_delete` value is not the equivalent of "never". Then disable the `can_delete` function while the import is running.

Reply:
@ximinez I considered working with `can_delete` but thought it simpler to just add a separate parameter (`waitForImport`) to the `if` statement in `SHAMapStoreImp::run()`. As it stands, do you think the approach is satisfactory? Right now I don't worry about disabling database rotation, only delaying it if removing ledgers would interfere with our import. I opted to work with the existing configuration rather than modifying it during the import.